flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DeclarativeSlotManager.java

```diff
@@ -0,0 +1,609 @@
+	private final Map<JobID, String> jobMasterTargetAddresses = new HashMap<>();
+	private final HashMap<SlotID, CompletableFuture<Acknowledge>> pendingSlotAllocationFutures;
+
+	/** ResourceManager's id. */
+	private ResourceManagerId resourceManagerId;
+
+	/** Executor for future callbacks which have to be "synchronized". */
+	private Executor mainThreadExecutor;
+
+	/** Callbacks for resource (de-)allocations. */
+	private ResourceActions resourceActions;
+
+	/** True iff the component has been started. */
+	private boolean started;
+
+	public DeclarativeSlotManager(
+			ScheduledExecutor scheduledExecutor,
+			SlotManagerConfiguration slotManagerConfiguration,
+			SlotManagerMetricGroup slotManagerMetricGroup,
+			ResourceTracker resourceTracker,
+			SlotTracker slotTracker) {
+
+		Preconditions.checkNotNull(slotManagerConfiguration);
+		this.taskManagerRequestTimeout = slotManagerConfiguration.getTaskManagerRequestTimeout();
+		this.slotManagerMetricGroup = Preconditions.checkNotNull(slotManagerMetricGroup);
+		this.resourceTracker = Preconditions.checkNotNull(resourceTracker);
+
+		pendingSlotAllocationFutures = new HashMap<>(16);
+
+		this.slotTracker = Preconditions.checkNotNull(slotTracker);
+		slotTracker.registerSlotStatusUpdateListener(createSlotStatusUpdateListener());
+
+		slotMatchingStrategy = slotManagerConfiguration.getSlotMatchingStrategy();
+
+		taskExecutorManagerFactory = (executor, resourceActions) -> new TaskExecutorManager(
+			slotManagerConfiguration.getDefaultWorkerResourceSpec(),
+			slotManagerConfiguration.getNumSlotsPerWorker(),
+			slotManagerConfiguration.getMaxSlotNum(),
+			slotManagerConfiguration.isWaitResultConsumedBeforeRelease(),
+			slotManagerConfiguration.getRedundantTaskManagerNum(),
+			slotManagerConfiguration.getTaskManagerTimeout(),
+			scheduledExecutor,
+			executor,
+			resourceActions);
+
+		resourceManagerId = null;
+		resourceActions = null;
+		mainThreadExecutor = null;
+		slotRequestTimeoutCheck = null;
+		taskExecutorManager = null;
+
+		started = false;
+	}
+
+	private SlotStatusUpdateListener createSlotStatusUpdateListener() {
+		return (taskManagerSlot, previous, current, jobId) -> {
+			if (previous == SlotState.PENDING) {
+				cancelAllocationFuture(taskManagerSlot.getSlotId());
+			}
+
+			if (current == SlotState.PENDING) {
+				resourceTracker.notifyAcquiredResource(jobId, taskManagerSlot.getResourceProfile());
+			}
+			if (current == SlotState.FREE) {
+				resourceTracker.notifyLostResource(jobId, taskManagerSlot.getResourceProfile());
+			}
+
+			if (current == SlotState.ALLOCATED) {
+				taskExecutorManager.occupySlot(taskManagerSlot.getInstanceId());
+			}
+			if (previous == SlotState.ALLOCATED && current == SlotState.FREE) {
+				taskExecutorManager.freeSlot(taskManagerSlot.getInstanceId());
+			}
+		};
+	}
+
+	private void cancelAllocationFuture(SlotID slotId) {
+		final CompletableFuture<Acknowledge> acknowledgeCompletableFuture = pendingSlotAllocationFutures.remove(slotId);
```

Comment:
I think `pendingSlotAllocationFutures` is never populated with any futures. What's the idea behind this structure?

Reply:
It is _supposed_ to be populated in `allocateSlot`. It is mostly there since the current SM also does such book-keeping (just in the TaskManagerSlot), but it could very well be unnecessary at this point.

Reply:
After all we aren't really doing anything on cancellation, although it might be useful for debugging purposes?

Reply:
Ah ok, I think I remember. The idea was to disable the callback on the response from the `TaskExecutor` when calling `requestSlot`.